gap/semigroups/semieunit.gi

```diff
@@ -0,0 +1,715 @@
+                  "the first argument must be a finite group,");
+  fi;
+
+  anti_act := function(pt, g)
+    return act(pt, g ^ -1);
+  end;
+
+  hom := ActionHomomorphism(G, DigraphVertices(X), anti_act);
+
+  if not IsSubgroup(AutomorphismGroup(X), Image(hom)) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the first argument (a group) must act by order ",
+                  "automorphisms on the third argument (a partial order ",
+                  "digraph),");
+  elif not IsPartialOrderDigraph(X) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the third argument must be a partial order digraph,");
+  fi;
+
+  # To check if the edges of y are a subset of the edges of x we need to apply
+  # the labeling of y's vertices to its edges.
+  y_edges := List(DigraphEdges(Y), edge -> [DigraphVertexLabels(Y)[edge[1]],
+                                            DigraphVertexLabels(Y)[edge[2]]]);
+
+  if not (IsSubset(DigraphVertexLabels(X), DigraphVertexLabels(Y)) and
+          IsSubset(DigraphEdges(X), y_edges)) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the fourth argument must be a subdigraph of the third ",
+                  "argument,");
+  elif not IsJoinSemilatticeDigraph(Y) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the fourth argument must be a join semilattice digraph,");
+  fi;
+
+  # Check condition M2 (check that Y is an order ideal of X.)
+  # TODO: implement IsOrderIdeal for a subset of a partial order digraph.
+  out_nbrs := OutNeighbors(X);
+  for x in DigraphVertexLabels(X) do
+    if not x in DigraphVertexLabels(Y) then
+      for y in DigraphVertexLabels(Y) do
+        if Intersection(out_nbrs[x], out_nbrs[y]) = out_nbrs[x] then
+          ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                        "condition M2 is not satisfied - see the ",
+                        "documentation for details,");
+        fi;
+      od;
+    fi;
+  od;
+
+  orbs := Orbits(Image(hom));
+
+  # Check condition M3 (check that G.Y = X .)
+  if not ForAll(orbs, o -> ForAny(DigraphVertexLabels(Y), v -> v in o)) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "condition M3 is not satisfied - see the documentation ",
+                  "for details,");
+  fi;
+
+  for x in DigraphVertexLabels(X) do
+    if not x in Union(orbs)
+        and not (x in DigraphVertexLabels(Y)) then
+      ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: \n",
+                    "Condition M3 is not satisfied - see the documentation ",
+                    "for details,");
+    fi;
+  od;
+
+  # Check condition M4 (essentially, check that X is connected.)
+  min := DigraphSinks(DigraphRemoveLoops(X))[1];
+  if ForAny(GeneratorsOfGroup(Image(hom)), g -> min ^ g <> min) then
```

Comment:
This should be the first sink of `Y` not of `X`. The only sink of `Y` should be the minimal element of the semilattice. 

M4: for all g in `G`,   The intersection of `gY` and `Y` is non-empty

Automorphisms of finite semilattices should fix the minimal element so checking this checks M4.

(Perhaps the comment about checking `X` is connected isn't very helpful. M3 checks that `X = GY` (i.e. all elements of X are images of elements of Y by G). So checking the minimal element of `Y` is fixed is equivalent to showing `X` is connected.)